### PR TITLE
Update 2D transform tutorial

### PR DESCRIPTION
**Pull request by Sauermann**
Update "Viewport and canvas transforms" tutorial

- incldue missing Window-transform
- Update graphic to make the distinction between coordinate systems
and transforms mode clear
- Restructure "transform functions" section
- Update functions to new Godot 4 conventions

Create a more detailed page in engine documentation about 2d coordinate
systems and 2d transforms.

changes this page: https://docs.godotengine.org/en/latest/tutorials/2d/2d_transforms.html
preview: https://github.com/Sauermann/godot-docs/blob/fix-transform-2d-update/tutorials/2d/2d_transforms.rst
new page preview: https://github.com/Sauermann/godot-docs/blob/fix-transform-2d-update/contributing/development/core_and_modules/2d_coordinate_systems.rst

Incorporates the changes from the recently merged godotengine/godot#59682, godotengine/godot#66688, godotengine/godot#66692 and godotengine/godot#68627.

### CODE REVIEW

**Comment by TheYellowArchitect** [comment]
While I cannot read the code accurately, if I understood correctly it's a great feature, and docs update alongside it is ofc much appreciated. Btw, please post the existing docs page, so its easier for everyone to see the diff in its rendered final form (e.g. sometimes images won't be rendered in added files, see .webp examples)

**Comment by Sauermann** [comment]
@TheYellowArchitect thank you for letting me know, that there is a functionality for previewing the page in its changed form. Added a link in the OP.

**Comment by TheYellowArchitect** [comment]
@Sauermann to explain the functionality preview change: Markdown renders differently (e.g. :ref:`Node3D <class_Node3D>` looks very different than the text format here on github, with a huge `REF` in front of the class name while rendered) so at least my workflow to review any docs request is to see the diff, then open the existing documentation page to confirm

I suggest editing the OP link with .html (I apologize for my first post that I posted .rst, I meant to post the .html which I have now edited :sweat_smile: )

**Comment by TheYellowArchitect** [comment]
That moment when a docs page adds many new sections instead of fixing bugs and typos, impressive :tada:

File: tutorials/2d/2d_transforms.rst [hunk 1]
```diff
@@ -1,84 +1,182 @@
...
-| Type                             | Transform                                                                                   |
-+==================================+=============================================================================================+
-| CanvasItem                       | :ref:`CanvasItem.get_global_transform() <class_CanvasItem_method_get_global_transform>`     |
-+----------------------------------+---------------------------------------------------------------------------------------------+
-| CanvasLayer                      | :ref:`CanvasItem.get_canvas_transform() <class_CanvasItem_method_get_canvas_transform>`     |
-+----------------------------------+---------------------------------------------------------------------------------------------+
-| CanvasLayer+GlobalCanvas+Stretch | :ref:`CanvasItem.get_viewport_transform() <class_CanvasItem_method_get_viewport_transform>` |
-+----------------------------------+---------------------------------------------------------------------------------------------+
+    canvasPos = GetGlobalTransform().Xform(localPos)
+    localPos = GetGlobalTransform().AffineInverse().Xform(canvasPos)
```
Comment (raulsntos) [inline tutorials/2d/2d_transforms.rst:167]: Transform2D also has a `*` operator in C#, the `Xform` method was removed.

```suggestion
    canvasPos = GetGlobalTransform() * localPos;
    localPos = GetGlobalTransform().AffineInverse() * canvasPos;
```

File: tutorials/2d/2d_transforms.rst [hunk 2]
```diff
@@ -1,84 +1,182 @@
...
 .. tabs::
  .. code-tab:: gdscript GDScript
 
-    var screen_coord = get_viewport_transform() * (get_global_transform() * local_pos)
+    var screen_coord = get_viewport().get_screen_transform() * get_global_transform_with_canvas() * local_pos
 
  .. code-tab:: csharp
 
-    var screenCord = (GetViewportTransform() * GetGlobalTransform()).Xform(localPos);
+    var screenCord = (GetViewport().GetScreenTransform() * GetGlobalTransformWithCanvas()).Xform(localPos);
```
Comment (raulsntos) [inline tutorials/2d/2d_transforms.rst:179]:
```suggestion
    var screenCord = GetViewport().GetScreenTransform() * GetGlobalTransformWithCanvas() * localPos;
```

File: tutorials/2d/2d_transforms.rst [hunk 3]
```diff
@@ -88,23 +186,22 @@ screen resolution resizing to work properly.
...
 
  .. code-tab:: csharp
 
     var localPos = new Vector2(10,20); // local to Control/Node2D
     var ie = new InputEventMouseButton();
-    ie.ButtonIndex = (int)ButtonList.Left;
-    ie.Position = (GetViewportTransform() * GetGlobalTransform()).Xform(localPos);
-    GetTree().InputEvent(ie);
+    ie.ButtonIndex = MouseButton.Left;
+    ie.Position = (GetViewport().GetScreenTransform() * GetGlobalTransformWithCanvas()).Xform(localPos);
```
Comment (raulsntos) [inline tutorials/2d/2d_transforms.rst:206]:
```suggestion
    ie.Position = GetViewport().GetScreenTransform() * GetGlobalTransformWithCanvas() * localPos;
```

File: tutorials/2d/2d_transforms.rst (hunk 1 above)
Comment (Sauermann) [reply tutorials/2d/2d_transforms.rst:167]: Thank you for the comments. I will include your improvements in the PR.

**Comment by mhilbrunner** [comment]
@Sauermann Once concern I have is that it replaces what currently is a very beginner-friendly docs page that answers basic questions with something very much more aimed at experts and that is more akin to something of detailed internal documentation of how everything works.

Both are useful. And we _did_ decide to move more to a technical manual than beginner tutorial for the official docs, it seems.
Still, I worry that replacing the old with the new will just make every non-expert mentally check out after seeing that new graphic. I know I did :) To compare:

# Old
old

# New
new


And a lot of that may be because of the increased complexity in Godot 4.0 with all its window handling, sure. The new page certainly contains a ton more useful information. It just seems way less accessible.

Maybe it would make sense to reorder the page a bit, to move the graphic down? Not sure.

Maybe some input from @NathanLovato could help.

**Comment by Sauermann** [comment]
Thanks for the review. It is an interesting idea to merge the previous beginner-friendly (but slightly wrong) explanation with the details of this PR. I am also open for ideas on how to improve this page.

**Comment by NathanLovato** [comment]
I'll give you a first answer, reviewing the general approach of the article, because we still have work for the 4.0 release.

The original article promises to discuss low-level details of how transforms work in the engine, so the amount of detail seems in line with this goal. But considering this is in the user manual, rather than engine documentation, I would personally change the target to giving users just what matters in their work.

I would move low-level details that are mostly required when contributing to the engine (i.e. knowledge that's not directly applicable when creating games in the editor) to the engine documentation.

Teaching-wise, the main thing I'm missing from the previous version and this PR are concrete takeaways or use cases for the reader in the intro. Something like:

> You will learn how ... You will be able to use that to do [insert concrete thing here] and [insert other example here].

And you want to make sure the following lecture reaches those goals.

This could be, for example, gaining the required knowledge to code a custom camera, if that's ever wanted.

The way I would approach teaching transforms in the user manual would be focusing a lot more on the fact that transform values pack position, rotation, and scale information, and how to read and manipulate those concretely, rather than covering the whole chain of transforms the engine applies to place and render images. And I'd cover the latter in engine documentation.

------

The image linked by @mhilbrunner is difficult to read. It tries to pack too much information, and an issue is that it has a lot of English text which won't be translatable or accessible to text-to-speech users. Images in the docs should generally not use more text than node types or symbols for this reason.

And again, it feels more like technical documentation for an engine contributor than what a user of the Godot editor would use in their projects.

In summary, my general feedback on this is that I don't think the goals and angle taken by the original article are the best fit for end user documentation. I feel that in its current state, it'd be a better fit for engine documentation.

**Comment by Sauermann** [comment]
I understand that there are some reservations against this PR in the end user documentation because of its complexity.

> I feel that in its current state, it'd be a better fit for engine documentation

Could I get some guidance about how to put the collected information and the graphic into the engine documentation?

**Comment by mhilbrunner** [comment]
@Sauermann Need to discuss this with the docs team, but currently, we got the `Contributing > Engine Development > Godot file formats` section, which has stuff like this: https://docs.godotengine.org/en/stable/contributing/development/file_formats/gdscript_grammar.html

It basically currently only has that and the TSCN format description, thats all. I could see renaming that "Godot File Formats" section to something like "Godot Engine Internals" and move this information there. Would be a good place to move a few other more internal/core development oriented information too. 

I also wonder if we should move that section one level higher. 🤔

**Comment by mhilbrunner** [comment]
After a brief discussion: https://docs.godotengine.org/en/stable/contributing/development/core_and_modules/index.html#getting-started-with-godot-s-source-code already has subpages for Core types, Object and Variant, so a page there for advanced Transform information would be a good fit.

So I guess to get this merged, we need:
1. To split it between the information important for the average beginner and Godot gamedev and more advanced/detailed info that goes into a page in https://docs.godotengine.org/en/stable/contributing/development/core_and_modules/index.html#getting-started-with-godot-s-source-code
2. Make another pass at making it a bit more accessible, as its a bit dense currently

**Comment by Sauermann** [comment]
I have updated the PR to split the details into a engine documentation page and leave most of the 2d-transform-tutorial intact with the exceptions of some updates.

**Review by RandomShaper — APPROVED** [review]
Looks great. This in an amazing piece of work.

My only other suggestion is to always use the full 'OS Window Manager' term, never leaving the "OS" part out. The reason is for extra clarity, as a skimmer or someone just checking a specific piece of the doc may otherwise come across the naked 'Window Manager' term and somehow think that it's some component of the engine.

**Comment by mhilbrunner** [comment]
Thanks for the review Pedro. @Sauermann Happy to merge this as soon as the above is addressed :)

File: contributing/development/core_and_modules/2d_coordinate_systems.rst [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    This is the local coordinate system of a :ref:`CanvasItem <class_CanvasItem>`.
+
+- **Parent Item Coordinates**
+    This is the local coordinate system of the parent's *CanvasItem*. When positioning
+    *CanvasItems* in the *Canvas*, they usually inherit the transformations of their parent
+    *CanvasItems*. An exceptions is
+    :ref:`CanvasItems.top_level <class_CanvasItem_property_top_level>`.
+
+- **Canvas Coordinates**
+    As mentioned in the previous tutorial :ref:`doc_canvas_layers`, there are two types of canvases
```
[the remark this answers is not in the document]
Comment (Sauermann) [reply contributing/development/core_and_modules/2d_coordinate_systems.rst:47]: Thanks for the comments, I have updated the PR to address your improvements.

**Comment by mhilbrunner** [comment]
@Sauermann Finally merged, wow this took a while :) All of this should be applicable for all of 4.0, 4.1 and 4.2+, right?

**Comment by Sauermann** [comment]
Hehe, doesn't beat my longest waiting PR ;-)
The relevant PRs got merged during 4.0 beta, so I would say yes.
